### DISCUSSION TITLE
Updating an instance buffer.

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [amerkoleci/Vortice.Windows#475] [Q&A] [answered]
Can anyone help with this buffer updating code? Not much too it, but it just doesn't work. The buffer initialization looks good and it's working without the update code. But the update code just tromps all over the buffer and set's it to mostly zero's with some odd values at the start.
```
  public TileData[] Tiles;
     
  ....
  
  // update instance buffer
  ID3D11DeviceContext dc = Gpu.DeviceContext;
  MappedSubresource mappedResource = dc.Map(InstanceBuffer, MapMode.WriteDiscard);
  Unsafe.Copy(mappedResource.DataPointer.ToPointer(), ref Tiles);
  dc.Unmap(InstanceBuffer, 0);
```
And the buffer desc...
```
  ReadOnlySpan<TileData> instanceDataSpan = new(Tiles);
  BufferDescription desc = new(
      (uint)(TileData.SizeInBytes * XTiles * YTiles),
      BindFlags.ShaderResource,
      ResourceUsage.Dynamic,
      CpuAccessFlags.Write,
      ResourceOptionFlags.BufferStructured,
      structureByteStride: TileData.SizeInBytes);
  
  InstanceBuffer?.Dispose();
  InstanceBuffer = device.CreateBuffer(instanceDataSpan, desc);
```

**Comment by Gavin-Williams** [comment]
The data in the buffer looks like it might be a pointer. That reminds me of some situations where data needs to be pinned. Or the ref to the array data is giving the pointer rather than the data. I tried creating a local reference.
> TileData[] tiles = Tiles;

But that didn't work. I succeeded by using Unsafe.CopyBlock instead of using the Copy method with the following code
```
  // update instance buffer
  ID3D11DeviceContext dc = Gpu.DeviceContext;
  MappedSubresource mappedResource = dc.Map(InstanceBuffer, MapMode.WriteDiscard);

  void* dstPtr = mappedResource.DataPointer.ToPointer();
  fixed (TileData* srcPtr = Tiles)
  {
      uint bytes = (uint)(sizeof(TileData) * XTiles * YTiles);
      Unsafe.CopyBlock(dstPtr, srcPtr, bytes);
  }
  dc.Unmap(InstanceBuffer, 0);
```
But I still can't get Copy to work.

**Reply by amerkoleci** [reply, marked as answer]
Hi,
Use one of the SetData helpers methods I've added in the pass:

https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.Direct3D11/ID3D11Buffer.cs#L10

**Reply by Gavin-Williams** [reply]
Ok, thank you, that makes it easy.

**Reply by amerkoleci** [reply]
> Ok, thank you, that makes it easy.

Thats why I'm here :)
